emailproxy.py

```diff
@@ -570,6 +570,79 @@ def _save_cache(cache_store_identifier, output_config_parser):
             Log.error('Error saving state to cache store file at', cache_store_identifier, '- is the file writable?')
 
 
+class Cryptographer:
+    ITERATIONS = 870000  # taken from cryptography's suggestion of using Django's defaults
+    LEGACY_ITERATIONS = 100000
+
+    def __init__(self, config, username, password):
+        """Creates a cryptographer which allows encrypting and decrypting sensitive information for this account,
+        including any stored tokens."""
+        self._salt = None
+
+        # try to base64 decode the existing salt
+        token_salt = config.get(username, 'token_salt', fallback=None)
+        if token_salt:
+            try:
+                self._salt = base64.b64decode(token_salt.encode('ascii'))  # catch incorrect third-party proxy guide
+            except (binascii.Error, UnicodeError):
+                Log.info('%s: Invalid `token_salt` value found in config file entry for account %s - this value is not '
+                         'intended to be manually created; generating new `token_salt`' % (APP_NAME, username))
+
+        # generate a new salt if the salt cannot be decoded or this is the initial run
+        if not self._salt:
+            self._salt = os.urandom(16)
+
+        # try to read the user configured token iterations
+        try:
+            iterations = int(config.get(username, 'token_iterations', fallback=self.LEGACY_ITERATIONS))
+        except ValueError:
+            iterations = self.LEGACY_ITERATIONS
+
+        # the first fernet is the primary fernet so sort the iterations count descending
+        self._iterations_options = sorted({self.ITERATIONS, iterations, self.LEGACY_ITERATIONS}, reverse=True)
```

Comment:
I'm curious – I'd always do `self.ITERATIONS + iterations + self.LEGACY_ITERATIONS` here. What's the reason for `{self.ITERATIONS, iterations, self.LEGACY_ITERATIONS}`?

Reply:
since it's a set it will drop out any duplicates since `iterations` _should_ be `self.ITERATIONS`

Reply:
Sorry, I had in my head that these values were lists after making the first comment above (and hence this would not work). Ignore this one!